File: syllabus/academia/views/gradRequirements.py
from syllabus import *

from syllabus.academia.models import Major, College, MajorRequirement, ClassProfile

def home(request):
    
    #majors = Major.objects.all().distinct('name')
    majors = Major.objects.all().order_by('name')
    colleges = College.objects.all()
    
    return render_to_response('/registrar/graduationRequirements/home.html', locals())

def viewCollegeDegree(request):
    
    college = College.objects.get(id = request.GET['id'])
    
    return render_to_response('/registrar/graduationRequirements/viewCollege.html', locals())

def viewDegree(request):
    
    major = Major.objects.get(id = request.GET['id'])
    
    if major.type != 'BS':
        bs = Major.objects.filter(name=major.name).filter(type='BS')
        if bs:
            major = bs[0]
    
    degrees = [major.type]
    
    for degree in Major.objects.filter(name=major.name).exclude(type=major.type):
        if major.type not in degrees:
            degrees.append(major.type)
            
    if type in request.GET:
        target = Major.objects.filter(name=major.name).filter(type=request.GET['type'])
        if target:
            major = target[0]
        
    
    preMajor = major.preMajor
    requirements = major.major
        
    return render_to_response('/registrar/graduationRequirements/viewDegree.html',locals())

def newCollegeRequirement(request):
    college = College.objects.get(id=request.GET['id'])
    interests = []
    for profile in ClassProfile.objects.all():
        if profile.interest not in interests:
            interests.append(profile.interest)
    
    
    return render_to_response('/registrar/graduationRequirements/requirementForm.html',locals())
    

def newRequirement(request):
    major = Major.objects.get(id=request.GET['major'])
    interests = []
    
    for profile in ClassProfile.objects.all():
        if profile.interest not in interests:
            interests.append(profile.interest)
    
    
    return render_to_response('/registrar/graduationRequirements/requirementForm.html',locals())


def editRequirement(request):
    requirement = MajorRequirement.objects.get(id=request.GET['id'])
    
    interests = []
    profiles = {}
    
    
    for profile in requirement.courses.all():
        if profile.interest not in profiles:
            profiles[profile.interest]=[]
            
        profiles[profile.interest] = ClassProfile.objects.filter(interest=profile.interest)
    
    for profile in ClassProfile.objects.all():
        if profile.interest not in interests:
            interests.append(profile.interest)
    
    
    return render_to_response('/registrar/graduationRequirements/requirementForm.html',locals())

def submitCollegeRequirement(request):
    
    requirement = MajorRequirement()
    requirement.name = request.POST['name']
    requirement.minGrade = request.POST['minGrade']
    requirement.number = request.POST['number']
    requirement.save()
    
    for id in request.POST['classes'].split(','):
        if id: 
            profile = ClassProfile.objects.get(id=id)
            if profile:
                requirement.courses.add(profile)
    
    College.objects.get(id=request.POST['id']).requirements.add(requirement)
    
    return HttpResponse('success!')

def majorList(request):
    
    majors = Major.objects.all().order_by('name')
    
    return render_to_response('/registrar/graduationRequirements/majorList.html', locals())

def newMajor(request):
    
    colleges = College.objects.all()
    
    return render_to_response('/registrar/graduationRequirements/majorForm.html', locals())

def createMajor(request):
    
    if College.objects.filter(id=request.POST['college']):
        major = Major()
        major.name = request.POST['name']
        major.type = request.POST['type']
        major.save()
        
        College.objects.get(id=request.POST['college']).majors.add(major)
        
        return HttpResponse('success!')

def submitRequirement(request):
    
    requirement = MajorRequirement()
    requirement.minGrade = request.POST['minGrade']
    requirement.number = request.POST['number']
    requirement.save()
    
    for id in request.POST['classes'].split(','):
        if id: 
            profile = ClassProfile.objects.get(id=id)
            if profile:
                requirement.courses.add(profile)
    
    major = Major.objects.get(id=request.POST['id'])
                
    if request.POST['level'] == 'preMajor':
        major.preMajor.add(requirement)
    else:
        major.major.add(requirement)
    
    return HttpResponse('success!')


def modifyRequirement(request):
    
    requirement = MajorRequirement.objects.get(id=request.POST['id'])
    requirement.minGrade = request.POST['minGrade']
    requirement.number = request.POST['number']
    requirement.save()
    
    requirement.courses.clear()
    
    for id in request.POST['classes'].split(','):
        if id: 
            profile = ClassProfile.objects.get(id=id)
            if profile:
                requirement.courses.add(profile)
                
    return HttpResponse('success!')

def profilesByInterest(request):
    
    interest = request.GET['interest']
    profiles = ClassProfile.objects.filter(interest__abbrv = interest)
    
    return render_to_response('/registrar/graduationRequirements/profileList.html',locals())
    
def deleteRequirement(request):
    
    requirement = MajorRequirement.objects.filter(id=request.GET['id'])
    
    requirement.delete()
    
    return HttpResponse('success!')
        
    
    
    

File: web/resources/scripts/registrar/gradRequirements.js
$(document).ready(function(){
    $('.filter').watermark('Search Majors...');
    $('input.filter').on('change',function(){
        if ($('.filter').val()){
            $('.major').show();
            $('.major').each(function(){
                if($(this).html().indexOf($('.filter').val()) == -1){
                    $(this).hide();    
                }

            });
        } else {
            $('.major').show();
        }
    });
    
    $('.major').on('click',function(){
        $('.current').removeClass('current');
        $(this).addClass('current');
    });
    $('.college').on('click',function(){
        $('.current').removeClass('current');
        $(this).addClass('current');
    });
});

function viewCollegeDegree(id){
    $.ajax({
        url:'/registrar/graduationRequirements/viewCollegeDegree/',
        data:{
            id: id    
        },
        success: function(data){
            $('#degree').empty().append(data);
        }
    });
}

function viewDegree(id, type){
    $.ajax({
        url: '/registrar/graduationRequirements/viewDegree/',
        data: {
            id: id,
            type: type
        },
        success: function(data) {
            $('#degree').empty().append(data);
            
            if (type){
                $('.degree').each(function(){
                    if ($(this).html() == type){
                        $(this).addClass('current');   
                    }
                });
            } else {
                $('.degree').eq(0).addClass('current');
            }
        }
    });
}

function newCollegeRequirement(id){
    
    $.ajax({
        url:'/registrar/graduationrequirements/newCollegeRequirement/',
        data:{
            id: id,
        },
        success: function(data) {
            overlay(data);
        }
    });
}

function newRequirement(id){
    
    $.ajax({
        url:'/registrar/graduationrequirements/newRequirement/',
        data:{
            major: id,
            type: $('.degree.current').html()
        },
        success: function(data) {
            overlay(data);
        }
    });
}

function editRequirement(id){
    $.ajax({
        url:'/registrar/graduationrequirements/editRequirement/',
        data:{
            id:id
        },
        success: function(data) {
            overlay(data);
        }
    });
}

function profilesByInterest(self, interest){
    
    $.ajax({
        url: '/registrar/graduationRequirements/profilesByInterest/',
        data: {
            interest: interest
        },
        success: function(data){
            $(self).siblings('#profileList').eq(0).empty().html(data);
        }
    });
}

function selectedProfile(){
    var id = $('.formLine').eq(-1).children('#profileList').val();
    if (id){
        if(Math.floor(id) == id && $.isNumeric(id)){
            var clone = $('.formLine:not(:last-child)').eq(-1).clone();  
            $('.formLine').eq(-1).after(clone);
            
            $(clone).children('#profileList').children(':not(:first-child)').remove();
        }
    }
}

function submitCollegeRequirement(idd){
    var name = $('.formLine').eq(0).children('input').val();
    var abbrv = $('.formLine').eq(1).children('input').val();
    var number = parseInt($('.formLine').eq(2).children('input').val());
    var minGrade = parseInt($('.formLine').eq(3).children('input').val());
    var classes = '';
    $('.formLine').not(':last-child').not(':first-child').each(function(){
        var id = $(this).children('#profileList').val();
        if (id){
            if( Math.floor(id) == id && $.isNumeric(id)){
                classes = classes + id + ','
            }
        }
    });
    

    $.ajax({
        url: '/registrar/graduationRequirements/submitCollegeRequirement/',
        type: 'POST',
        data: {
            number : number,
            classes : classes,
            minGrade : minGrade,
            name: name,
            id: idd,
        },
        success: function(data){
            closeOverlay();
            viewCollegeDegree(idd, $('.degree.current').html());
        }
    });
}

function submitRequirement(idd){
    var number = parseInt($('.formLine').eq(0).children('input').val());
    var minGrade = parseInt($('.formLine').eq(1).children('input').val());
    var classes = '';
    $('.formLine').not(':last-child').not(':first-child').each(function(){
        var id = $(this).children('#profileList').val();
        if (id){
            if( Math.floor(id) == id && $.isNumeric(id)){
                classes = classes + id + ','
            }
        }
    });
    

    $.ajax({
        url: '/registrar/graduationRequirements/submitRequirement/',
        type: 'POST',
        data: {
            number : number,
            classes : classes,
            minGrade : minGrade,
            id: idd,
            level: $("input:checked").val()
        },
        success: function(data){
            closeOverlay();
            viewDegree(idd, $('.degree.current').html());
        }
    });
}

function updateMajors(){
    $.ajax({
        url: '/registrar/graduationRequirements/majorList/',
        success: function(data) {
            $('.major').remove();
            $('.filter').parent().after(data);
        }
    });
}

function createMajor() {
    $.ajax({
        url: '/registrar/graduationRequirements/createMajor/',
        type: 'POST',
        data: {
            college: $('#college').val(),
            name: $('#name').val(),
            type: $('#type').val()
        },
        success: function(data){
            updateMajors();
            closeOverlay();
        }
    });
}

function newMajor(){
    $.ajax({
        url: '/registrar/graduationRequirements/newMajor/',
        success: function(data){
            overlay(data);
        }
    });
}

function deleteRequirement(id){
    if (confirm('Are you sure you want to delete this requirement?')){
        $.ajax({
        url: '/registrar/graduationrequirements/deleteRequirement',
        data:{
            id: id
        },
        success: function(data){
            if ($('.degree.current').length > 0){
                $('.degree.current').click();
            } else {
                $('.college.current').click();
            };
        }
    });
    }
}

function modifyRequirement(idd){
    var number = parseInt($('.formLine').eq(0).children('input').val());
    var minGrade = parseInt($('.formLine').eq(1).children('input').val());
    var classes = '';
    $('.formLine').not(':last-child').not(':first-child').each(function(){
        var id = $(this).children('#profileList').val();
        if (id){
            if( Math.floor(id) == id && $.isNumeric(id)){
                classes = classes + id + ','
            }
        }
    });
    

    $.ajax({
        url: '/registrar/graduationRequirements/modifyRequirement/',
        type: 'POST',
        data: {
            number : number,
            classes : classes,
            minGrade : minGrade,
            id: idd,
            level: $("input:checked").val()
        },
        success: function(data){
            closeOverlay();
            viewDegree(idd, $('.degree.current').html());
        }
    });
}